### PR TITLE
[WIP] Postcodes v0

### PR DESCRIPTION
**Pull request by batinic-m**
# Description

Adding postcode type to Addresses theme in order to have an encompassing schema which incorporates all data types. The initial drop of postcodes is intended to contain [GeoNames](https://download.geonames.org/export/zip/) data covering EU countries, since the coverage there is of the best quality. This gives us a good starting point for future sidecars and expansion into other countries.

# Testing

Testing of the schema and the examples/counterexamples done via the git-inferred tests done during push.

# Checklist

*Checklist of tasks commonly-associated with schema pull requests. Please review the relevant checklists and ensure you do all the tasks that are required for the change you made.*

1. [x] Add relevant examples.
2. [x] Add relevant counterexamples.
3. [x] Update any counterexamples that became obsolete. For example, if a counterexample uses property `A` but is not intended to test property `A`'s validity, and you made a schema change that invalidates property `A` in that counterexample, fix the counterexample to align it with your schema change.  
4. [x] Update in-schema documentation using plain English written in complete sentences, if an update is required.
5. [ ] Update Docusaurus documentation, if an update is required.
6. [ ] Review change with Overture technical writer to ensure any advanced documentation needs will be taken care of, unless the change is trivial and would not affect the documentation.

# Documentation Website

[Docs preview for this PR.](https://dfhx9f55j8eg5.cloudfront.net/pr/223)

### CODE REVIEW

File: schema/addresses/defs.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        within it.
+        
+        Not every country has to have each granularity of the postcode. 
+        Some countries have only one layer of postcodes, and some don't even use postcodes, like Angola, Bahamas, Botswana etc.
+      type: string
+      enum:
+        - postcode1       # Most granular, most detailed set of postcodes in a country.
+                          # E.g. in the UK: KT 10 9RT
+                          # And in Germany: 30657
+
```
Comment (jwass) [inline schema/addresses/defs.yaml:20]: Are these phrases, "first-order" "second-order" postcodes commonly known? Could be - I'm just not familiar with that phrase.

File: schema/addresses/postcode.yaml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      - "$ref": https://geojson.org/schema/MultiPolygon.json
+  properties:   # GeoJSON: top-level object 'properties' property.
+    unevaluatedProperties: false
+    allOf:
+      - "$ref": ../defs.yaml#/$defs/propertyContainers/overtureFeaturePropertiesContainer
+    required: [name]
+    properties: # JSON Schema: properties within GeoJSON top-level object 'properties' property
+      name:
+         pattern: |  
+            if country == 'AT' and subtype == 'postcode1' then ^\d{4}$
```
Comment (jwass) [inline schema/addresses/postcode.yaml:41]: Small nit: but can you put these in alphabetical order?

File: schema/addresses/postcode.yaml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            elif country == 'PL' and subtype == 'postcode1' then ^\d{2}-\d{3}$
+            elif country == 'PT' and subtype == 'postcode1' then ^\d{4}-\d{3}$
+            elif country == 'RO' and subtype == 'postcode1' then ^\d{6}$
+            elif country == 'SK' and subtype == 'postcode1' then ^\d{3}\s?\d{2}$
+            elif country == 'SI' and subtype == 'postcode1' then ^\d{4}$
+            elif country == 'ES' and subtype == 'postcode1' then ^\d{5}$
+            elif country == 'SE' and subtype == 'postcode1' then ^\d{3}\s?\d{2}$
+      subtype: { "$ref": "./defs.yaml#/$defs/typeDefinitions/postcodetype" }
+      country: { "$ref": "../defs.yaml#/$defs/propertyDefinitions/iso3166_1Alpha2CountryCode" }
+      city_division_id:
```
Comment (jwass) [inline schema/addresses/postcode.yaml:70]: "city" isn't a concept in divisions schema. Does this field always point to a "locality"?

https://github.com/OvertureMaps/schema/blob/main/schema/divisions/defs.yaml#L12

File: schema/addresses/postcode.yaml (hunk 2 above)
Comment (jwass) [inline schema/addresses/postcode.yaml:41]: Are patterns known for postcode 2-4?

File: schema/addresses/postcode.yaml (hunk 3 above)
Comment (vcschapp) [reply schema/addresses/postcode.yaml]: I thought we agreed we weren't going to go ahead and prematurely try linking addresses to divisions.

File: schema/addresses/postcode.yaml (hunk 2 above)
Comment (vcschapp) [reply schema/addresses/postcode.yaml:41]: Is `if`/`then` inside patterns a known feature of JSON Schema?

Can you link to where it is defined? I'm not clear how a JSON Schema validator would distinguish between an `if`/`then` statement and a regular expression.

File: counterexamples/addresses/bad-postcode-city-division-id.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  type: postcode
+  update_time: '2024-06-19T16:27:01-08:00'
+  version: 0
+  name: 67017
+  country: IT
+  subtype: postcode1
+  locality_division_id: " "
+  ext_expected_errors:
+    - "[I#/properties/locality_division_id] [S#/$defs/propertyDefinitions/id/pattern] does not match pattern"
+
```
Comment (vcschapp) [inline counterexamples/addresses/bad-postcode-city-division-id.yaml:18]: Please fix your editor to add newlines at the end of files.

**Review by vcschapp — CHANGES_REQUESTED** [review]
There are 111 commits in the commit list as of this morning! This PR probably needs to be rebased and resubmitted.

**Comment by jwass** [comment]
It looks like we're holding off on this to focus on address core functions. We can revive it later if necessary

**Pull request closed without merging**